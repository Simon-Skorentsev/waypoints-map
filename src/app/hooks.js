import { useState } from "react";

export const useWpName = (initialState) => {
    let name = initialState;
    const setName = (newName) => {
        name = newName;
    }
    return [name, setName];
}