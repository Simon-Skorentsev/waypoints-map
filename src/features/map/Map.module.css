.map {
	width: 900px;
	height: 60vh;
}
