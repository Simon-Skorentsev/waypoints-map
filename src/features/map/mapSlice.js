import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  waypoints: [],  //[{"coords": [0, 0], "id": 0, "name": "first"}]
};

export const mapSlice = createSlice({
  name: 'Map',
  initialState,
  reducers: {
    addWaypoint(state, action) {
      state.waypoints.push(action.payload);
    },

    deleteWaypoint(state, action) {
      state.waypoints.splice(action.payload, 1);
    },

    dragWaypoint(state, action) {
      state.waypoints.forEach(waypoint => {
        if (waypoint.id === action.payload.id) {
          waypoint.coords = action.payload.coords;
        };
      });
    },

    replaceWaypointsList(state, action) {
      state.waypoints = action.payload;
    }
  }
});

export const { initMap, addWaypoint, dragWaypoint, deleteWaypoint, replaceWaypointsList } = mapSlice.actions;

export default mapSlice.reducer;
