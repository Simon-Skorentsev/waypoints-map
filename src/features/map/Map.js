import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styles from './Map.module.css';
import { addWaypoint, deleteWaypoint, dragWaypoint } from './mapSlice';
import { v4 as uuid } from 'uuid';
import { store } from '../../app/store';
import { Waypoints } from '../waypoints/Waypoints';

export function Map() {
  const dispatch = useDispatch();
  const { waypoints } = useSelector(state => state.mapSlice);
  const [map, setMap] = useState();
  const [polyline, setPolyline] = useState();
  const [waypointNumber, setWaypointNumber] = useState(1);
  const [wpCollection, setWpCollection] = useState();
  const ymaps = window.ymaps;
  
  const getWaypoints = () => store.getState().mapSlice.waypoints;
  
  const updatePolilyne = (numInState, actualCoords) => {
    const newAllCoords = polyline.geometry.getCoordinates().map((oldCoords, i) => {
      return i === numInState ? actualCoords : oldCoords
    })
    polyline.geometry.setCoordinates(newAllCoords);
  }

  const createMap = (containerId = "map", settings = { center: [55.76, 37.64], zoom: 10 }) => {
    const { center, zoom } = settings;

    const myMap = new ymaps.Map(containerId, {
      center,
      zoom
    });

    const polyline = new ymaps.Polyline([]);  //маршрут между точками на карте
    const wpCollection = new ymaps.GeoObjectCollection();  //необходимо для удаления меток с карты

    myMap.geoObjects.add(polyline);
    myMap.geoObjects.add(wpCollection);

    setMap(myMap);
    setPolyline(polyline);
    setWpCollection(wpCollection);
  };

  //создание метки в центре карты
  const addMapWaypoint = (name = waypointNumber) => {
    const center = map.getCenter(),
      id = uuid();
    const placemark = new ymaps.Placemark([...center], {}, {
      balloonPanelMaxMapArea: 0,
      draggable: "true",
      preset: "islands#blueStretchyIcon",
      openEmptyBalloon: true
    });

    placemark.properties.set("balloonContentHeader", name);

    wpCollection.add(placemark);
    dispatch(addWaypoint(({ "coords": [...center], id, name })));
    
    placemark.events.add(
      "drag", () => {
        const actualCoords = placemark.geometry.getCoordinates(),
          actualWaypoints = getWaypoints();

        const numInState = actualWaypoints.findIndex(waypoint => (
          waypoint.id === id
        ));

        dispatch(dragWaypoint({"coords": [...actualCoords], id}));
        
        updatePolilyne(numInState, actualCoords);
      }
    );

    placemark.events.add(  //показ адреса при клике на балун
      "click", () => {
        placemark.properties.set("balloonContent", "поиск...");

        ymaps.geocode(placemark.geometry.getCoordinates()).then(function (res) {
          const firstGeoObject = res.geoObjects.get(0);

          placemark.properties.set({
            // Формируем строку с данными об объекте.
            iconCaption: [
              // Название населенного пункта или вышестоящее административно-территориальное образование.
              firstGeoObject.getLocalities().length
                ? firstGeoObject.getLocalities()
                : firstGeoObject.getAdministrativeAreas(),
              // Получаем путь до топонима, если метод вернул null, запрашиваем наименование здания.
              firstGeoObject.getThoroughfare() || firstGeoObject.getPremise(),
            ]
              .filter(Boolean)
              .join(", "),
            // В качестве контента балуна задаем строку с адресом объекта.
            balloonContent: firstGeoObject.getAddressLine(),
          });
        })
      }
    )

    polyline.geometry.setCoordinates([...polyline.geometry.getCoordinates(), center,]);  //встаиваем точку в маршрут
    setWaypointNumber(waypointNumber + 1);
  };

  const deleteMapWaypoint = (id) => {
    const numInState = waypoints.findIndex(waypoint => (
      waypoint.id === id
    ));

    polyline.geometry.remove(numInState);
    wpCollection.splice(numInState, 1);
    dispatch(deleteWaypoint(numInState));
  }

  useEffect(() => {
    createMap();
  }, [])

  return (
    <>
      <div id="map" className={styles.map} />
      <Waypoints addMapWaypoint={addMapWaypoint} 
      deleteMapWaypoint={deleteMapWaypoint} 
      updatePolilyne={updatePolilyne}
      wpCollection={wpCollection}
      />
    </>
  );
};