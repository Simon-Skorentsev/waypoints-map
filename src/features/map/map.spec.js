import { Map } from "./map";
import renderer from 'react-test-renderer';
import { Provider } from "react-redux";
import { store } from "../../app/store";

test("shoud render waypointsList and waypoints input", () => {
    const el = renderer.create(
        <Provider store={store}>
            <Map/>
        </Provider>
    );
    
    expect(el).toMatchSnapshot();
});