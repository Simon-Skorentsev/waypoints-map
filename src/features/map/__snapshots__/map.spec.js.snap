// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shoud render waypointsList and waypoints input 1`] = `
Array [
  <div
    className="map"
    id="map"
  />,
  <div
    className="waypoints"
  >
    <div
      className="MuiFormControl-root MuiTextField-root wpInput css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      onKeyDown={[Function]}
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
      >
        Введите название метки
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-173wfxe"
          >
            <span>
              Введите название метки
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <ul
      className="MuiList-root MuiList-padding waypointsList css-h4y409-MuiList-root"
    />
  </div>,
]
`;
